### PR TITLE
feat(Autocomplete): Add infinite scroll autocomplete

### PR DESCRIPTION
**Pull request by Thebarda**
single
multi

### CODE REVIEW

File: src/InputField/Select/Autocomplete/InfiniteScroll/index.tsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      }
+
+      if (optionsOpen) {
+        loadOptions({
+          endpoint: getEndpoint({ search: searchValue, page }),
+          loadMore: page > 1,
+        });
+      } else if (!optionsOpen) {
+        setOptions([]);
+        setPage(initialPage);
```
Comment (bdauria) [inline src/InputField/Select/Autocomplete/InfiniteScroll/index.tsx:125]: I'm not sure this block will ever be reachable

File: src/utils/useObserver.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      }
+      if (loading) {
+        observer.current = null;
+      } else {
+        observer.current = new IntersectionObserver((entries) => {
+          if (entries[0].isIntersecting && page < maxPage) {
+            action();
+          }
+        });
+      }
```
Comment (bdauria) [inline src/utils/useObserver.ts:30]:
```suggestion
        observer.current = null;
        return;
      }
      observer.current = new IntersectionObserver((entries) => {
        if (entries[0].isIntersecting && page < maxPage) {
          action();
        }
      });
      
```

File: src/InputField/Select/Autocomplete/InfiniteScroll/index.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        return;
+      }
+
+      if (optionsOpen) {
+        loadOptions({
+          endpoint: getEndpoint({ search: searchValue, page }),
+          loadMore: page > 1,
+        });
+      }
```
Comment (bdauria) [inline src/InputField/Select/Autocomplete/InfiniteScroll/index.tsx:123]:
```suggestion
      loadOptions({
        endpoint: getEndpoint({ search: searchValue, page }),
        loadMore: page > 1,
      });
```

**Comment by Thebarda** [comment]
> Is it possible to add a test on infinite scroll ? (loading of page 2)

We can not test that because Jest tests on generated code not but not graphically like Cypress.

**Comment by kduret** [comment]
Ok, event if I'm not totally sure this should be validated with Cypress :)
for example, material-ui validate scroll on tabs component : https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tabs/Tabs.test.js#L371

**Comment by Thebarda** [comment]
> Ok, even if I'm not totally sure this should be validated with Cypress :)
> for example, material-ui validate scroll on tabs component : https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tabs/Tabs.test.js#L371

Ok I will do that later

**Comment by Thebarda** [comment]
> > Ok, even if I'm not totally sure this should be validated with Cypress :)
> > for example, material-ui validate scroll on tabs component : https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tabs/Tabs.test.js#L371
> 
> Ok I will do that later even if I have no idea how to do that...

@kduret I looked out a little to Tabs implementation about scroll event. 
It seems they can manage scrollable tabs : https://material-ui.com/components/tabs/#scrollable-tabs
With a `onScroll` listener : https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Tabs/Tabs.js#L488
With that they can fully control scroll state.
In our case, scroll bar is uncontrolled. We only check if last element appears on screen when scrolling. 
Both uses cases are completly different. 
Furthermore, we mock IntersectionObserver for our tests. So.... IntersectionObserver is not triggered.

During the implementation, I tried to test that feature but without success because of these reasons

**Comment by kduret** [comment]
Ok, thanks for investigation ! this is much clearer now :)

**Comment by jdelpierre** [comment]
I think we should post a loader for the slightly long response times. 
We should show that we have finished loading everything (end of the number of results).

**Pull request merged**